File: generators/humanoid/gltf_pipeline/skeleton.py
"""Skeleton definitions for the humanoid gltf_pipeline.

19-bone hierarchy with rest positions, parent indices, and inverse bind matrices.
All bone positions are computed proportionally based on character height H.
"""

import numpy as np

# ── Bone list (index order matches JOINTS in glTF skin) ──────────────────────

BONE_NAMES = [
    "Hips",                                             # 0  — root
    "Spine", "Chest", "Neck", "Head",                  # 1-4
    "Shoulder.L", "UpperArm.L", "LowerArm.L", "Hand.L",  # 5-8
    "Shoulder.R", "UpperArm.R", "LowerArm.R", "Hand.R",  # 9-12
    "UpperLeg.L", "LowerLeg.L", "Foot.L",              # 13-15
    "UpperLeg.R", "LowerLeg.R", "Foot.R",              # 16-18
]

NUM_BONES = len(BONE_NAMES)  # 19

# ── Parent indices (-1 = root) ────────────────────────────────────────────────

BONE_PARENTS = [
    -1,   # Hips
     0,   # Spine → Hips
     1,   # Chest → Spine
     2,   # Neck → Chest
     3,   # Head → Neck
     2,   # Shoulder.L → Chest
     5,   # UpperArm.L → Shoulder.L
     6,   # LowerArm.L → UpperArm.L
     7,   # Hand.L → LowerArm.L
     2,   # Shoulder.R → Chest
     9,   # UpperArm.R → Shoulder.R
    10,   # LowerArm.R → UpperArm.R
    11,   # Hand.R → LowerArm.R
     0,   # UpperLeg.L → Hips
    13,   # LowerLeg.L → UpperLeg.L
    14,   # Foot.L → LowerLeg.L
     0,   # UpperLeg.R → Hips
    16,   # LowerLeg.R → UpperLeg.R
    17,   # Foot.R → LowerLeg.R
]


def compute_bone_world_positions(H: float) -> np.ndarray:
    """Compute rest (bind) world positions for each bone proportional to height H.

    Args:
        H: Character height in meters.

    Returns:
        Array of shape (19, 3) float32 with (x, y, z) world positions.
    """
    positions = np.array([
        # Spine / head chain
        [0.0,       0.0,       H * 0.52],   # 0  Hips
        [0.0,       0.0,       H * 0.60],   # 1  Spine
        [0.0,       0.0,       H * 0.68],   # 2  Chest
        [0.0,       0.0,       H * 0.82],   # 3  Neck
        [0.0,       0.0,       H * 0.87],   # 4  Head
        # Left arm chain
        [+H * 0.08, 0.0,       H * 0.72],   # 5  Shoulder.L
        [+H * 0.14, 0.0,       H * 0.70],   # 6  UpperArm.L
        [+H * 0.14, 0.0,       H * 0.53],   # 7  LowerArm.L
        [+H * 0.14, 0.0,       H * 0.38],   # 8  Hand.L
        # Right arm chain
        [-H * 0.08, 0.0,       H * 0.72],   # 9  Shoulder.R
        [-H * 0.14, 0.0,       H * 0.70],   # 10 UpperArm.R
        [-H * 0.14, 0.0,       H * 0.53],   # 11 LowerArm.R
        [-H * 0.14, 0.0,       H * 0.38],   # 12 Hand.R
        # Left leg chain
        [+H * 0.09, 0.0,       H * 0.50],   # 13 UpperLeg.L
        [+H * 0.09, 0.0,       H * 0.27],   # 14 LowerLeg.L
        [+H * 0.09, H * 0.08,  H * 0.03],   # 15 Foot.L
        # Right leg chain
        [-H * 0.09, 0.0,       H * 0.50],   # 16 UpperLeg.R
        [-H * 0.09, 0.0,       H * 0.27],   # 17 LowerLeg.R
        [-H * 0.09, H * 0.08,  H * 0.03],   # 18 Foot.R
    ], dtype=np.float32)
    return positions


def compute_inverse_bind_matrices(world_positions: np.ndarray) -> np.ndarray:
    """Compute inverse bind matrices for each bone.

    Since bones are just translations at rest (no rotation), the bind matrix is:
        [[1, 0, 0, -px],
         [0, 1, 0, -py],
         [0, 0, 1, -pz],
         [0, 0, 0,   1]]

    glTF stores matrices in column-major order.

    Args:
        world_positions: Array of shape (19, 3).

    Returns:
        Array of shape (19, 4, 4) float32 — inverse bind matrices in column-major
        layout (ready to flatten with .flatten(order='F') or .T.flatten() per bone).
    """
    n = len(world_positions)
    inv_bind = np.zeros((n, 4, 4), dtype=np.float32)

    for i, (px, py, pz) in enumerate(world_positions):
        # Row-major inverse bind matrix (translation negated)
        mat = np.eye(4, dtype=np.float32)
        mat[0, 3] = -px
        mat[1, 3] = -py
        mat[2, 3] = -pz
        inv_bind[i] = mat

    return inv_bind


def bone_name_to_index(name: str) -> int:
    """Return the bone index for a given bone name, or -1 if not found."""
    try:
        return BONE_NAMES.index(name)
    except ValueError:
        return -1

File: js/src/skeleton.js
/**
 * 19-bone skeleton hierarchy using BABYLON.Bone/Skeleton.
 * Port of generators/humanoid/gltf_pipeline/skeleton.py
 */

import { Skeleton, Bone, Matrix, TransformNode } from '@babylonjs/core';

export const BONE_NAMES = [
  'Hips',                                              // 0  — root
  'Spine', 'Chest', 'Neck', 'Head',                   // 1-4
  'Shoulder.L', 'UpperArm.L', 'LowerArm.L', 'Hand.L', // 5-8
  'Shoulder.R', 'UpperArm.R', 'LowerArm.R', 'Hand.R', // 9-12
  'UpperLeg.L', 'LowerLeg.L', 'Foot.L',               // 13-15
  'UpperLeg.R', 'LowerLeg.R', 'Foot.R',               // 16-18
];

export const BONE_PARENTS = [-1,0,1,2,3,2,5,6,7,2,9,10,11,0,13,14,0,16,17];

/**
 * Return world positions for each bone proportional to height H.
 * @param {number} H - character height in metres
 * @returns {Array<[number,number,number]>} array of [x,y,z] per bone
 */
export function boneWorldPositions(H) {
  // [x, y, z] — Y = height (up), Z = forward depth, X = lateral
  return [
    // Spine/head chain
    [0.0,          H * 0.52, 0.0        ],  // 0  Hips
    [0.0,          H * 0.60, 0.0        ],  // 1  Spine
    [0.0,          H * 0.68, 0.0        ],  // 2  Chest
    [0.0,          H * 0.82, 0.0        ],  // 3  Neck
    [0.0,          H * 0.87, 0.0        ],  // 4  Head
    // Left arm chain
    [+H * 0.08,    H * 0.72, 0.0        ],  // 5  Shoulder.L
    [+H * 0.14,    H * 0.70, 0.0        ],  // 6  UpperArm.L
    [+H * 0.14,    H * 0.53, 0.0        ],  // 7  LowerArm.L
    [+H * 0.14,    H * 0.38, 0.0        ],  // 8  Hand.L
    // Right arm chain
    [-H * 0.08,    H * 0.72, 0.0        ],  // 9  Shoulder.R
    [-H * 0.14,    H * 0.70, 0.0        ],  // 10 UpperArm.R
    [-H * 0.14,    H * 0.53, 0.0        ],  // 11 LowerArm.R
    [-H * 0.14,    H * 0.38, 0.0        ],  // 12 Hand.R
    // Left leg chain
    [+H * 0.09,    H * 0.50, 0.0        ],  // 13 UpperLeg.L
    [+H * 0.09,    H * 0.27, 0.0        ],  // 14 LowerLeg.L
    [+H * 0.09,    H * 0.03, H * 0.08  ],  // 15 Foot.L (Z = toe forward)
    // Right leg chain
    [-H * 0.09,    H * 0.50, 0.0        ],  // 16 UpperLeg.R
    [-H * 0.09,    H * 0.27, 0.0        ],  // 17 LowerLeg.R
    [-H * 0.09,    H * 0.03, H * 0.08  ],  // 18 Foot.R (Z = toe forward)
  ];
}

/**
 * Build a BABYLON.Skeleton with 19 bones at rest positions proportional to H.
 * @param {number} H - character height in metres
 * @param {BABYLON.Scene} scene - Babylon scene the skeleton belongs to
 * @returns {BABYLON.Skeleton}
 */
/**
 * Build a BABYLON.Skeleton with 19 bones at rest positions proportional to H.
 * Each bone is linked to a TransformNode so the GLTF exporter can resolve the
 * bone's local transform and export animation tracks correctly.
 *
 * @param {number} H - character height in metres
 * @param {BABYLON.Scene} scene
 * @returns {{ skeleton: BABYLON.Skeleton, boneNodes: Map<string, BABYLON.TransformNode> }}
 */
export function buildSkeleton(H, scene) {
  const worldPos = boneWorldPositions(H);
  const skeleton  = new Skeleton('Skeleton', 'Skeleton', scene);

  const bones = [];
  const nodes = [];                // parallel array: TransformNode per bone
  const boneNodes = new Map();     // boneName → TransformNode

  for (let i = 0; i < BONE_NAMES.length; i++) {
    const name      = BONE_NAMES[i];
    const parentIdx = BONE_PARENTS[i];
    const parentBone = parentIdx === -1 ? null : bones[parentIdx];
    const [wx, wy, wz] = worldPos[i];

    // Local offset from parent (or world pos for root)
    let lx, ly, lz;
    if (parentIdx === -1) {
      lx = wx; ly = wy; lz = wz;
    } else {
      const [px, py, pz] = worldPos[parentIdx];
      lx = wx - px; ly = wy - py; lz = wz - pz;
    }

    // Bone with rest-pose matrix
    const localMatrix = Matrix.Translation(lx, ly, lz);
    const bone = new Bone(name, skeleton, parentBone, localMatrix);
    bones.push(bone);

    // TransformNode linked to this bone — required for GLTF export
    const tn = new TransformNode(name, scene);
    if (parentIdx !== -1) tn.parent = nodes[parentIdx];
    tn.position.set(lx, ly, lz);
    bone.linkTransformNode(tn);

    nodes.push(tn);
    boneNodes.set(name, tn);
  }

  return { skeleton, boneNodes };
}
